### DISCUSSION TITLE
How to generate Table type instead of class type

### DISCUSSION THREAD

**Discussion by phpsxg** [agronholm/sqlacodegen#173] [Q&A]
When using sqlacodegen to automatically generate a model from clickhouse, the generated model is of Table type. How to make the automatically generated model be of class
**class**

```# coding: utf-8
from sqlalchemy import BIGINT, Column, Float, INTEGER, Index, SMALLINT, TIMESTAMP, VARCHAR, text
from sqlalchemy.dialects.mysql.types import TINYINT
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()
metadata = Base.metadata


class SqlacodegenTest(Base):
    __tablename__ = 'sqlacodegen_test'
    __table_args__ = (
        Index('idx_house_agent_site', 'code', 'site', unique=True),
    )

    id = Column(BIGINT(20), primary_key=True)
    city_id = Column(INTEGER(11), nullable=False, server_default=text("'0'"))
    site = Column(SMALLINT(6), nullable=False, server_default=text("'1'"))
    code = Column(VARCHAR(20), nullable=False, index=True)
    type = Column(INTEGER(11), nullable=False, server_default=text("'0'"))
    quality = Column(Float(asdecimal=True), nullable=False, server_default=text("'0'"))
    rank = Column(INTEGER(11), nullable=False, server_default=text("'1'"))
    detail = Column(VARCHAR(500), nullable=False, server_default=text("'0'"))
    status = Column(TINYINT(4), nullable=False, server_default=text("'0'"))
    ctime = Column(TIMESTAMP, nullable=False, server_default=text("'0000-00-00 00:00:00'"))
    mtime = Column(TIMESTAMP, nullable=False, server_default=text("CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP"))
```

```# coding: utf-8
from sqlalchemy import BIGINT, Column, Float, INTEGER, Index, MetaData, SMALLINT, TIMESTAMP, Table, VARCHAR, text
from sqlalchemy.dialects.mysql.types import TINYINT

metadata = MetaData()


t_sqlacodegen_test = Table(
    'sqlacodegen_test', metadata,
    Column('id', BIGINT(20), nullable=False),
    Column('city_id', INTEGER(11), nullable=False, server_default=text("'0'")),
    Column('site', SMALLINT(6), nullable=False, server_default=text("'1'")),
    Column('code', VARCHAR(20), nullable=False, index=True),
    Column('type', INTEGER(11), nullable=False, server_default=text("'0'")),
    Column('quality', Float(asdecimal=True), nullable=False, server_default=text("'0'")),
    Column('rank', INTEGER(11), nullable=False, server_default=text("'1'")),
    Column('detail', VARCHAR(500), nullable=False, server_default=text("'0'")),
    Column('status', TINYINT(4), nullable=False, server_default=text("'0'")),
    Column('ctime', TIMESTAMP, nullable=False, server_default=text("'0000-00-00 00:00:00'")),
    Column('mtime', TIMESTAMP, nullable=False, server_default=text("CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP")),
    Index('idx_house_agent_site', 'code', 'site', unique=True)
)
```

**Comment by agronholm** [comment]
SQLAlchemy model classes require primary keys. If no primary key constraint is found, a `Table` is generated instead.

**Reply by phpsxg** [reply]
> SQLAlchemy model classes require primary keys. If no primary key constraint is found, a `Table` is generated instead.

There is a primary key, clickhouse sort field will be used as the primary key. 
For example, this table:
```
CREATE TABLE default.student
(

    `id` Int64,

    `age` Int8,

    `name` String
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192


```
The primary_key field in the system.tables is equal to `id`

**Reply by agronholm** [reply]
If the column is not marked as the primary key, sqlacodgen cannot know that.

**Reply by phpsxg** [reply]
> If the column is not marked as the primary key, sqlacodgen cannot know that.

My table building statement is as follows, the export able modle is still Table,  **clickhouse database**
CREATE TABLE default.student
(

    `id` Int64,

    `age` Int8,

    `name` String
)
ENGINE = MergeTree
PRIMARY KEY id
ORDER BY id
SETTINGS index_granularity = 8192
```
```

**Reply by agronholm** [reply]
Can you paste the generated table here (the one with a primary key)?

**Reply by agronholm** [reply]
Is this on MySQL or one of its derivatives?

**Reply by phpsxg** [reply]
> Can you paste the generated table here (the one with a primary key)?

```
# coding: utf-8
from sqlalchemy import Column, MetaData, String, Table
from clickhouse_sqlalchemy.types.common import Int64, Int8

metadata = MetaData()


t_student = Table(
    'student', metadata,
    Column('id', Int64),
    Column('age', Int8),
    Column('name', String)
)

```

**Reply by phpsxg** [reply]
> Is this on MySQL or one of its derivatives?

ClickHouse®is a column-oriented database management system (DBMS) for online analytical processing of queries (OLAP).
https://clickhouse.com/docs/en/

**Reply by agronholm** [reply]
It looks like reflection does not properly detect primary keys with this driver/backend combination. This needs to be fixed upstream, either in SQLAlchemy or ClickHouse itself. If the reflected metadata does not contain primary keys for the table, it will generate a `Table` instead of a class. There is no support in sqlacodegen to force classes to be generated in the absence of primary keys.

**Comment by phpsxg** [comment]
@agronholm
 I have another question, some of the automatically generated columns have classes, how to make the automatically generated ones without classes, thanks

**clickhouse database**
```

t_DIM_INFO = Table(
    'DIM_INFO', metadata,
    Column('DIM_TYPE_CODE', String),
    Column('DIM_TYPE_NAME', Nullable(<class 'clickhouse_sqlalchemy.types.common.String'>)),
    Column('DIM_CODE', String),
    Column('DIM_NAME', Nullable(<class 'clickhouse_sqlalchemy.types.common.String'>)),
    Column('DIM_VALUE_CODE', String),
    Column('DIM_VALUE_NAME', Nullable(<class 'clickhouse_sqlalchemy.types.common.String'>)),
    Column('DIM_FORMAT_DESC', Nullable(<class 'clickhouse_sqlalchemy.types.common.String'>)),
    Column('DIM_MEMO', Nullable(<class 'clickhouse_sqlalchemy.types.common.String'>)),
    Column('_sign', Int8),
    Column('_version', UInt64)
)

```

**Reply by agronholm** [reply]
I don't know how that can happen, but then again I've only ever tried this against relational databases. Have you tried with the latest beta (3.0.0b2)?

**Reply by phpsxg** [reply]
> 

The same is true for version 3.0.0b2. **ClickHouse database**

**Reply by agronholm** [reply]
I will try to reproduce this on my own using `clickhouse_sqlalchemy`.

**Comment by shl-hsn** [comment]
I have encountered the same problem. 
I use `clickhouse_sqlalchemy ` to generate table but not class. 
Have you completed it?
